ethereum_history_api/circuits/lib/src/log.nr

```diff
@@ -1,22 +1,127 @@
-use crate::misc::types::{Address, Bytes32};
+use crate::{
+    misc::{
+    types::{Address, Bytes32, ADDRESS_LENGTH, BYTES32_LENGTH}, fragment::Fragment,
+    arrays::subarray_inferred_len
+},
+    rlp::{decode::decode_list, types::{RlpList, RlpFragment}},
+    verifiers::receipt::{MAX_RECEIPT_RLP_LENGTH, rlp::RECEIPT_FIELDS_COUNT}
+};
 use dep::std::unsafe::zeroed;
 
+global MAX_TOPICS_COUNT = 4;
+global LOG_FIELDS_COUNT = 3;
+global LOGS_INDEX = 3;
+
+global LOG_ADDRESS_INDEX = 0;
+global LOG_TOPICS_INDEX = 1;
+global LOG_DATA_INDEX = 2;
+
+struct PhantomMaxLogDataLen<MAX_LOG_DATA_LEN> {}
+struct PhantomMaxLogsCount<MAX_LOGS_COUNT> {}
+
 struct Log<MAX_LOG_DATA_LEN> {
     address: Address,
-    topics: BoundedVec<Bytes32, 4>,
+    topics: BoundedVec<Bytes32, MAX_TOPICS_COUNT>,
     data: BoundedVec<u8, MAX_LOG_DATA_LEN>,
 }
 
+impl<MAX_LOG_DATA_LEN> Eq for Log<MAX_LOG_DATA_LEN> {
+    fn eq(self, other: Self) -> bool {
+        (self.address == other.address) &
+        (self.topics == other.topics) &
+        (self.data == other.data)
+    }
+}
+
 struct LogWithinBlock<MAX_LOG_DATA_LEN> {
     log: Log<MAX_LOG_DATA_LEN>,
     block_hash: Bytes32
 }
 
-pub fn get_log<MAX_LOG_DATA_LEN>(
-    _block_number: Field,
-    _tx_idx: Field,
-    _log_idx: Field
-) -> LogWithinBlock<MAX_LOG_DATA_LEN> {
+struct LogFieldsRlpFragments {
+    address: RlpFragment,
+    topics: BoundedVec<RlpFragment, MAX_TOPICS_COUNT>,
+    data: RlpFragment
+}
+
+pub fn get_log<MAX_LOG_DATA_LEN>(_block_number: Field, _tx_idx: Field, _log_idx: Field) -> LogWithinBlock<MAX_LOG_DATA_LEN> {
     assert(false, "Not implemented");
     zeroed()
 }
+
+pub(crate) fn extract_log<MAX_LOG_DATA_LEN, MAX_LOGS_COUNT>(
+    receipt_rlp: [u8; MAX_RECEIPT_RLP_LENGTH],
+    log_idx: u64,
+    phantom_max_log_data_len: PhantomMaxLogDataLen<MAX_LOG_DATA_LEN>,
+    phantom_max_logs_count: PhantomMaxLogsCount<MAX_LOGS_COUNT>
+) -> Log<MAX_LOG_DATA_LEN> {
+    let log_fragment = extract_log_fragment(receipt_rlp, log_idx, phantom_max_logs_count);
+    let log_fields_rlp_fragments = extract_log_fields_fragments(log_fragment);
+    convert_to_log(
+        receipt_rlp,
+        log_fields_rlp_fragments,
+        phantom_max_log_data_len
+    )
+}
+
+fn extract_log_fragment<MAX_LOGS_COUNT>(
```

Comment:
This function does two separate things. Let's split it into two functions. That would allow us to reuse variable names